File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        nums_dict = {}
        for x in nums:
            if x in nums_dict:
                nums_dict[x] += 1
            else:
                nums_dict[x] = 1
        min_heap = []
        for num in nums_dict:
            heappush(min_heap, (nums_dict[num], num))
            if len(min_heap) > k:
                heappop(min_heap)
        freq_ele = []
        for tup in min_heap:
            freq, ele = tup
            freq_ele.append(ele)
        return freq_ele

        
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Freq {
    int ele, freq;

    public Freq(int ele, int freq){
        this.ele = ele;
        this.freq = freq;
    }

    public int getFreq(){
        return freq;
    }

    public int getEle(){
        return ele;
    }
}

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> freq = new HashMap<>();
        for(int n : nums){
            if(freq.containsKey(n)){
                freq.put(n, freq.get(n)+1);
            }
            else{
                freq.put(n, 1);
            }
        }
        PriorityQueue<Freq> min_heap = new PriorityQueue<>((a, b) -> a.getFreq() - b.getFreq());

        freq.forEach((Integer a, Integer b) -> {
            Freq obj = new Freq(a, b);
            min_heap.add(obj);
            if(min_heap.size() > k){ min_heap.poll(); }
        });

        int[] output = new int[k];
        for(int i=0; i<k; i++){
            output[i] = min_heap.poll().getEle();
        }
        return output;
    }
}